File: sdk/blocktypes/chunks.go
package blocktypes

import (
	"crypto/sha512"
	"encoding/hex"
	"errors"
	"fmt"

	"github.com/bpprotocol/blockparty/sdk/blockpb"
)

var (
	// ErrMissingChunk is returned when a manifest references a chunk that is not
	// available for reassembly.
	ErrMissingChunk = errors.New("blocktypes: missing chunk")
	// ErrChecksum is returned when reassembled bytes do not match the manifest's
	// SHA-512.
	ErrChecksum = errors.New("blocktypes: reassembled checksum mismatch")
)

// ReassembleChunks concatenates the chunk payloads named by chunkIDs (in order)
// and verifies the result against sha512Hex (the manifest's hex SHA-512).
// dataByID maps a chunk block ID to that chunk's payload bytes.
func ReassembleChunks(chunkIDs []string, dataByID map[string][]byte, sha512Hex string) ([]byte, error) {
	var buf []byte
	for _, id := range chunkIDs {
		d, ok := dataByID[id]
		if !ok {
			return nil, fmt.Errorf("%w: %s", ErrMissingChunk, id)
		}
		buf = append(buf, d...)
	}
	sum := sha512.Sum512(buf)
	if hex.EncodeToString(sum[:]) != sha512Hex {
		return nil, ErrChecksum
	}
	return buf, nil
}

// ReassembleChunkManifest reassembles a transport-layer chunk.manifest.
func ReassembleChunkManifest(m *blockpb.ChunkManifest, dataByID map[string][]byte) ([]byte, error) {
	return ReassembleChunks(m.Chunks, dataByID, m.Sha512)
}

// ReassembleContentChunked reassembles an application-layer
// content.chunked.manifest (its headers are handled separately by the caller).
func ReassembleContentChunked(m *blockpb.ContentChunkedManifest, dataByID map[string][]byte) ([]byte, error) {
	return ReassembleChunks(m.Chunks, dataByID, m.Sha512)
}

File: implementations/typescript/packages/sdk/src/blocktypes/chunks.ts
import { sha512 } from "../crypto/index.js";
import { bytesToHex, concatBytes } from "../bytes.js";
import type { ChunkManifest, ContentChunkedManifest } from "../blockpb/block_types_pb.js";

export class MissingChunkError extends Error {
  constructor(id: string) {
    super(`blocktypes: missing chunk: ${id}`);
    this.name = "MissingChunkError";
  }
}

export class ChecksumError extends Error {
  constructor() {
    super("blocktypes: reassembled checksum mismatch");
    this.name = "ChecksumError";
  }
}

/**
 * Concatenate the chunk payloads named by chunkIDs (in order) and verify the
 * result against sha512Hex (the manifest's hex SHA-512). `dataByID` maps a chunk
 * block ID to that chunk's payload bytes.
 */
export function reassembleChunks(
  chunkIDs: string[],
  dataByID: Map<string, Uint8Array>,
  sha512Hex: string,
): Uint8Array {
  const parts: Uint8Array[] = [];
  for (const id of chunkIDs) {
    const d = dataByID.get(id);
    if (!d) throw new MissingChunkError(id);
    parts.push(d);
  }
  const buf = concatBytes(...parts);
  if (bytesToHex(sha512(buf)) !== sha512Hex) throw new ChecksumError();
  return buf;
}

/** Reassemble a transport-layer chunk.manifest. */
export function reassembleChunkManifest(
  manifest: ChunkManifest,
  dataByID: Map<string, Uint8Array>,
): Uint8Array {
  return reassembleChunks(manifest.chunks, dataByID, manifest.sha512);
}

/** Reassemble an application-layer content.chunked.manifest. */
export function reassembleContentChunked(
  manifest: ContentChunkedManifest,
  dataByID: Map<string, Uint8Array>,
): Uint8Array {
  return reassembleChunks(manifest.chunks, dataByID, manifest.sha512);
}
